### PR TITLE
code server web: have a commit in the static and callback routes

### PR DESCRIPTION
**Pull request by aeschli**
The bult-in web server used for code server web accesses sources and artifacts using the `/static` and `/callback` routes.

The pr adds a `quality-commit` segment in from of these.

That helps with caching and allows to hand over from one server (running an older version) to another server, running the latest version.

- [x] update expiry date
- [ ] websoccket
- [ ] /vscode-remote-resource

### CODE REVIEW

File: src/vs/server/node/webClientServer.ts [hunk 1]
```diff
@@ -135,11 +139,11 @@ export class WebClientServer {
...
+		// Strip the this._staticRoute from the path
 		const normalizedPathname = decodeURIComponent(parsedUrl.pathname!); // support paths that are uri-encoded (e.g. spaces => %20)
-		const relativeFilePath = normalize(normalizedPathname.substr('/static/'.length));
+		const relativeFilePath = normalizedPathname.substring(this._staticRoute.length + 1);
 
-		const filePath = join(APP_ROOT, relativeFilePath);
+		const filePath = join(APP_ROOT, relativeFilePath); // join also normalizes the path
 		if (!isEqualOrParent(filePath, APP_ROOT, !isLinux)) {
 			return serveError(req, res, 400, `Bad request.`);
 		}
```
Comment (alexdima) [inline src/vs/server/node/webClientServer.ts:149]: We could add here something like:
```
if (this._productService.quality && this._productService.commit) {
	headers['Cache-Control'] = 'public, max-age=31536000';
}
```

**Review by alexdima — APPROVED** [review]

File: src/vs/server/node/webClientServer.ts [hunk 2]
```diff
@@ -148,7 +159,7 @@ export class WebClientServer {
...
 
-		return serveFile(this._logService, req, res, filePath, headers);
+		return serveFile(filePath, this._environmentService.isBuilt ? CacheControl.NO_EXPIRY : CacheControl.NO_CACHING, this._logService, req, res, headers);
```
Comment (alexdima) [inline src/vs/server/node/webClientServer.ts:162]: When running from source, this should be `CacheControl.ETAG` to maintain existing behaviour.

File: src/vs/server/node/webClientServer.ts [hunk 3]
```diff
@@ -78,14 +90,14 @@ export async function serveFile(logService: ILogService, req: http.IncomingMessa
 }
 
 const APP_ROOT = dirname(FileAccess.asFileUri('', require).fsPath);
-const CHARCODE_SLASH = '/'.charCodeAt(0);
+const CHARCODE_SLASH = CharCode.Slash;
```
Comment (alexdima) [inline src/vs/server/node/webClientServer.ts:93]: This could just be inlined instead of being assigned to a constant.

**Review by alexdima — APPROVED** [review]

File: src/vs/server/node/webClientServer.ts (hunk 2 above)
Comment (aeschli) [reply src/vs/server/node/webClientServer.ts:162]: When running from sources I prefer no caching and I don't see a benefit of sticking to ETags. So unless you strongly object, I'd make that change.

**Comment by aeschli** [comment]
Merged the web server changes. I'll create a new PR for the remaining changes.

File: src/vs/server/node/webClientServer.ts (hunk 2 above)
Comment (alexdima) [reply src/vs/server/node/webClientServer.ts:161]: Let's continue in https://github.com/microsoft/vscode/issues/148642

**Pull request merged**